packages/react/src/components/createInlineOverlayComponent.tsx

```diff
@@ -54,65 +55,43 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
-       * handler if present as setState will
-       * cause the event handlers to be
-       * destroyed and re-created.
-       */
-      this.ref.current?.addEventListener('willPresent', (evt: any) => {
-        this.setState({ isOpen: true });
+      this.ref.current?.addEventListener('ionMount', this.handleIonMount);
+      this.ref.current?.addEventListener('willPresent', this.handleWillPresent);
+      this.ref.current?.addEventListener('didDismiss', this.handleDidDismiss);
+    }
 
-        this.props.onWillPresent && this.props.onWillPresent(evt);
-      });
+    componentDidUpdate(prevProps: IonicReactInternalProps<PropType>) {
+      const node = this.ref.current! as HTMLElement;
+      attachProps(node, this.props, prevProps);
+    }
 
+    componentWillUnmount() {
+      const node = this.ref.current;
       /**
-       * Unmount the inner component.
-       * React will call Node.removeChild
-       * which expects the child to be
-       * a direct descendent of the parent
-       * but due to the presence of
-       * Web Component slots, this is not
-       * always the case. To work around this
-       * we move the inner component to the root
-       * of the Web Component so React can
-       * cleanup properly.
+       * If the overlay is being unmounted, but is still
+       * open, this means the unmount was triggered outside
+       * of the overlay being dismissed.
+       *
+       * This can happen with:
+       * - The parent component being unmounted
+       * - The overlay being conditionally rendered
+       * - A route change (push/pop/replace)
+       *
+       * Unmounting the overlay at this stage should skip
+       * the dismiss lifecycle, including skipping the transition.
+       *
        */
-      this.ref.current?.addEventListener('didDismiss', (evt: any) => {
-        const wrapper = this.wrapperRef.current;
-        const el = this.ref.current;
-
+      if (node && this.state.isOpen) {
         /**
-         * This component might be unmounted already, if the containing
-         * element was removed while the popover was still open. (For
-         * example, if an item contains an inline popover with a button
-         * that removes the item.)
+         * Detach the local event listener than performs the state updates,
+         * before dismissing the overlay, to prevent the callback handlers
+         * executing after the component has been unmounted. This is to
+         * avoid memory leaks.
          */
-        if (wrapper && el) {
-          el.append(wrapper);
-          this.setState({ isOpen: false });
-        }
-
-        this.props.onDidDismiss && this.props.onDidDismiss(evt);
-      });
-    }
-
-    componentDidUpdate(prevProps: IonicReactInternalProps<PropType>) {
-      const node = this.ref.current! as HTMLElement;
-      attachProps(node, this.props, prevProps);
+        node.removeEventListener('didDismiss', this.handleDidDismiss);
+        node.remove();
```

Comment:
Why do we need to manually remove the DOM node? Shouldn't React handle this for us?

Reply:
At this point in the execution/logic, the overlay is presented and relocated to the root of the DOM. It is outside of the context of React's VDOM to unmount. 

We could try teleporting the node back to have React also remove it, but we would need to also repaint the overlay as dismissed, otherwise it would cause visual overlapping issues with moving a presented overlay back into the nested DOM tree. 

I opted for `node.remove()` since it cleans up the presented overlay and doesn't require any style recalculations or visual changes. We unfortunately are dealing with a limited execution window with `componentWillUnmount`, so we cannot do a blocking task like `await node.dismiss()`.

Reply:
oh that's right. thanks!